logbook-json/src/main/java/org/zalando/logbook/json/JsonHttpLogFormatter.java

```diff
@@ -1,92 +1,184 @@
+    
+            generator.writeEndObject();
         }
-        if (JsonMediaType.JSON.test(contentType)) {
-        	// TODO has this JSON been validated? If not then this might result in invalid log statements
-            return Optional.of(new JsonBody(body));
-        } else {
-            return Optional.of(body);
+        
+        return writer.toString();
+            
+    }
+
+    private void writeBody(HttpMessage message, String body, JsonGenerator builder) throws IOException {
+        if(!body.isEmpty()) {
+            builder.writeFieldName("body");
+
+            final String contentType = message.getContentType();
+
+            if (JsonMediaType.JSON.test(contentType)) {
+                builder.writeRawValue(body);
+            } else {
+                builder.writeString(body);
+            }
         }
     }
 
     @Override
-    public String format(final Map<String, Object> content) throws IOException {
-        return mapper.writeValueAsString(content);
+    public String format(Correlation correlation, HttpResponse response) throws IOException {
+        
+        final String correlationId = correlation.getId();
+        
+        String body = response.getBodyAsString();
+
+        StringBuilderWriter writer = new StringBuilderWriter(body.length() + 2048);
+
+        try (JsonGenerator generator = jsonFactory.createGenerator(writer)) {
+            generator.writeStartObject();
+    
+            generator.writeFieldName("origin");
+            if(response.getOrigin() == Origin.LOCAL) {
+                generator.writeRawValue("\"local\"");
+            } else {
+                generator.writeRawValue("\"remote\"");
+            }
+            generator.writeFieldName("type");
+            generator.writeRawValue("\"response\"");
+            generator.writeStringField("correlation", correlationId);
+            generator.writeStringField("protocol", response.getProtocolVersion());
+            generator.writeNumberField("duration", correlation.getDuration().toMillis());
+            generator.writeNumberField("status", response.getStatus());
+    
+            writeHeaders(generator, response);
+            writeBody(response, body, generator);
+            
+            generator.writeEndObject();
+        }
+        return writer.toString();
+    }
+
+    protected void writeHeaders(JsonGenerator builder, HttpMessage httpMessage) throws IOException {
+        Map<String, List<String>> headers = httpMessage.getHeaders();
+        
+        if(!headers.isEmpty()) {
+            // implementation note:
+            // for some unclear reason, manually iterating over the headers
+            // while writing performs worse than letting Jackson do the job.
+            builder.writeObjectField("headers", headers);
+        }
     }
+    
+    private void reconstructUri(final HttpRequest request, JsonGenerator generator, StringBuilderWriter writer) throws IOException {
```

Comment:
Passing two *sinks* here is a bit strange.

Reply:
I can't remember how much of a difference the current reconstructUri(..) version did, so I'll give it a go later and compare against just doing this 'the normal way'.